onnx/defs/quantization/defs.cc

```diff
@@ -15,6 +15,68 @@ For saturation, it saturates to [0, 255] if it's uint8, or [-128, 127] if it's i
 For (x / y_scale), it's rounding to nearest ties to even. Refer to https://en.wikipedia.org/wiki/Rounding for details. 'y_zero_point' and 'y' must have same type.
 )DOC";
 
+bool BuildContextDependentFunctionBodyQuantizeLinear(
+    const FunctionBodyBuildContext& ctx,
+    const OpSchema& schema,
+    FunctionProto& functionProto) {
+  FunctionBuilder builder(functionProto);
+
+  auto* input_type = ctx.getInputType(0);
+  if (!input_type->has_tensor_type()) {
+    fail_schema("Invalid x input type. Expected a tensor type");
+  }
+  auto input_elt_type = input_type->tensor_type().elem_type();
+
+  auto output_elt_type = TensorProto::UINT8;
+  float quantized_min = 0, quantized_max = 255;
+
+  if (ctx.hasInput(2)) {
+    auto* zero_point_type = ctx.getInputType(2);
+    if (!zero_point_type->has_tensor_type()) {
+      fail_schema("Invalid y_zero_point input type. Expected a tensor type");
+    }
+    if (zero_point_type->tensor_type().elem_type() == TensorProto::UINT8) {
+      quantized_min = 0;
+      quantized_max = 255;
+      output_elt_type = TensorProto::UINT8;
+    } else if (zero_point_type->tensor_type().elem_type() == TensorProto::INT8) {
+      quantized_min = -128;
+      quantized_max = 127;
+      output_elt_type = TensorProto::INT8;
+    } else {
+      fail_schema("Invalid zero_point_type. Expected TensorProto::UINT8 or TensorProto::INT8");
+    }
+  }
+
+  bool is_per_axis = ctx.getInputType(1)->tensor_type().shape().dim().size() == 1 &&
```

Comment:
I don't think we can rely on the static-type to figure out if this is per-axis or not: in general, the shape-information may not be available statically. Same problem for Dequantize function too.

Reply:
Both quantized element type and y_scale shape depends on the context. We could manage not to use y_scale shape (for example, to force broadcasting regardless, which will be inefficient in cases where broadcast is not needed). We can only get quantized element type from the context. Thus, the op function has to be context dependent.

Reply:
Agree. we can not fully rely on static shape inference